firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,301 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                    return Ok(Some((node.into_node(), value)));
+                }
+
+                let children = get_children_iter(branch);
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+        }
+
+        first_loop = false;
     }
 }
 
-/// Compute a key from a set of parents
-fn key_from_parents(parents: &[(ObjRef, u8)]) -> Vec<u8> {
-    parents[1..]
-        .chunks_exact(2)
-        .map(|parents| (parents[0].1 << 4) + parents[1].1)
-        .collect::<Vec<u8>>()
+fn get_children_iter(branch: &BranchNode) -> impl Iterator<Item = (DiskAddress, u8)> {
+    branch
+        .children
+        .into_iter()
+        .enumerate()
+        .filter_map(|(pos, child_addr)| child_addr.map(|child_addr| (child_addr, pos as u8)))
+}
+
+#[must_use]
+struct MustUse<T>(T);
+
+impl<T> From<T> for MustUse<T> {
+    fn from(t: T) -> Self {
+        Self(t)
+    }
+}
+
+impl<T: Not> Not for MustUse<T> {
+    type Output = T::Output;
+
+    fn not(self) -> Self::Output {
+        self.0.not()
+    }
+}
+
+fn next_child<'a, S, T, Iter>(
+    merkle: &'a Merkle<S, T>,
+    mut children: Iter,
+    parents: &mut Vec<(ObjRef<'a>, u8)>,
+    node: &mut ParentNode<'a>,
+    pos: &mut u8,
+) -> Result<MustUse<bool>, super::MerkleError>
+where
+    Iter: Iterator<Item = (DiskAddress, u8)>,
+    S: ShaleStore<Node>,
+{
+    if let Some((child_addr, child_pos)) = children.next() {
+        let child = merkle.get_node(child_addr)?;
+
+        *pos = child_pos;
+        let node = std::mem::replace(node, ParentNode::New(child));
+        parents.push((node.into_node(), *pos));
+    } else {
+        let Some((next_parent, next_pos)) = parents.pop() else {
+            return Ok(false.into());
+        };
+
+        *node = ParentNode::Visited(next_parent);
+        *pos = next_pos;
+    }
+
+    Ok(true.into())
 }
-fn key_from_parents_and_leaf(parents: &[(ObjRef, u8)], leaf: &LeafNode) -> Vec<u8> {
-    let mut iter = parents[1..]
+
+/// create an iterator over the key-nibbles from all parents _excluding_ the sentinal node.
+fn nibble_iter_from_parents<'a>(parents: &'a [(ObjRef, u8)]) -> impl Iterator<Item = u8> + 'a {
+    parents
         .iter()
-        .map(|parent| parent.1)
-        .chain(leaf.path.to_vec());
-    let mut data = Vec::with_capacity(iter.size_hint().0);
-    while let (Some(hi), Some(lo)) = (iter.next(), iter.next()) {
+        .skip(1) // always skip the sentinal node
+        .flat_map(|(parent, child_nibble)| match parent.inner() {
+            NodeType::Branch(_) => Either::Left(std::iter::once(*child_nibble)),
+            NodeType::Extension(extension) => Either::Right(extension.path.iter().copied()),
+            NodeType::Leaf(leaf) => Either::Right(leaf.path.iter().copied()),
+        })
+}
+
+enum Either<T, U> {
```

Comment:
I have a similar set of questions to the previous set around `MustUse`; I'll avoid repeating them here.

I read that `Either` was removed from `std` during rust development because it was thought that the overly generic enums of `Left` and `Right` produced less readable code. For those users that still wanted it, they moved it over to the `either` crate.

Q1: Why did you choose to reimplement `Either` here rather than use the `either` crate?

Reply:
The `either` crate comes with a bunch of baggage that we don't need. I would argue that if we find more use cases for it other than this one place, then we should pull it in, but I don't think we need to pull in a whole crate just to get a little stack-based dynamic-dispatch that can be implemented in fewer than 20-lines of code.

Reply:
Fair. What are your thoughts about making it easy to remove later by perhaps sticking it in it's own module (can even be in the same source file)?

Reply:
If you want me to do that, I can, but I really don't see the benefit. All you have to do is delete the type.

Reply:
Your choice. For each comment, here are your options:

- You disagree with the review comment, in which case you should state why and resolve the comment.
- You agree with the review comment but it's not stopping the commit because you'll fix it later. If you don't have a follow-on commit pending, you'll create an issue.
- You agree with the review comment and will fix it before landing.
- There's a comment that refers to style or preference or a suggestion that doesn't matter to the reviewer very much (a "nit"). You can choose to ignore these, or treat them like one of the other 3. I will sometimes ignore them if they are marked nit.

If you don't see the benefit, then you've met the first bullet and you should resolve this comment.

Reply:
I moved the `Either` type to a module